### ISSUE TITLE
avahi.org/download has no file index

### ISSUE THREAD

**Issue by lathiat** [avahi/avahi.github.io#2]
Unlike the previous host, avahi.org/download has no index of files.

Some distributions such as Debian (using debian's debian/watch) file rely on this for their automatic checking of new releases.

We may need to fix this (it's possible to generate such an index with GitHub actions) and/or formalise a move to using GitHub releases or similar.

I know Debian does this, I am not sure who if anyone else does this.

**Comment by evverx** [comment]
I think it should be possible for Debian to switch to GitHub by analogy with https://salsa.debian.org/systemd-team/systemd/-/blob/debian/master/debian/watch. @mbiebl I wonder if it would be possible to switch https://salsa.debian.org/utopia-team/avahi/-/blob/debian/master/debian/watch from avahi.org?

**Comment by lathiat** [comment]
Yeah they'd totally be up to switch I imagine, I guess we just need to decide if we want to do that or not.

**Comment by evverx** [comment]
To judge from https://salsa.debian.org/utopia-team/avahi/-/commit/893a2769f18f4c444a186872a329bc5272e9598b Debian switched to GitHub and then switched back to avahi.org due to I assume that bug where GitHub messed up checksums. That change was reverted by GitHub so it should probably be safe to switch back.

I think GitHub releases should be good enough. If someone really needs the download folder they can implement and maintain that action themselves.

**Comment by mbiebl** [comment]
https://wiki.debian.org/debian/watch#GitHub has some background

I.e. the download links being JS based and one approach is to use api.github.com. The problem is, that access to api.github.com is rate limited.

**Comment by mbiebl** [comment]
That said, yes, we can adjust debian/watch to use github directly (whether it be releases or tags).

Just curious: Is it still planned to release full `make dist` tarballs or only git tag snapshots?

**Comment by evverx** [comment]
Personally I'd switch to snapshots (and maybe signed tags eventually) because it's easier to maintain. That being said if I used tarballs I wouldn't be happy about this because I don't trust snapshots generated by GitHub :-) I guess it depends on https://github.com/avahi/avahi/issues/449.

**Comment by mbiebl** [comment]
Yeah, unfortunately the Debian repository is still tarball based and I'm aware of the shortcomings of git archive / github snapshots.
So, I would prefer official release tarballs, but if it's too much effort, then Debian will certainly cope.

**Comment by mbiebl** [comment]
Btw, nice to see the @avahi project being in place now!

**Comment by mbiebl** [comment]
I staged those changes for now:
```
diff --git a/debian/watch b/debian/watch
index fbfcf52d..f0c0dce5 100644
--- a/debian/watch
+++ b/debian/watch
@@ -1,2 +1,4 @@
 version=4
-https://avahi.org/download/@PACKAGE@@ANY_VERSION@@ARCHIVE_EXT@
+opts="searchmode=plain" \
+  https://api.github.com/repos/avahi/@PACKAGE@/releases \
+  https://github.com/avahi/@PACKAGE@/releases/download/v(?:\d[\d.]*)/@PACKAGE@@ANY_VERSION@@ARCHIVE_EXT@
```
This assumes that there will still be releases going forward.

**Comment by lathiat** [comment]
I think publishing release tarballs is still fine, you can upload them to the GitHub release no problem. Is there a best practice other projects have moved towards to continue or avoid doing that?

If we wanted to stop there'd be a bit of work to do I think since the make dist generates some outputs such as teh docs, autoconf, etc.

**Comment by evverx** [comment]
> Is there a best practice other projects have moved towards to continue or avoid doing that?

I think it depends on the project. Some projects like systemd release git snapshots but it isn't exactly friendly to distros consuming tarballs (partly because those autogenerated snapshots can't be signed).

FWIW I agree that it would be better to keep releasing "make dist" tarballs.

(I don't think they should be kept in two different places though in the sense that I don't think the "download" folder at avahi.org is needed any more)

**Comment by mbiebl** [comment]
Let me clarify, that Debian does not depend on pre-generated files in the (dist) tarball.

What we would like to have though is a stable reference to the upstream tarball (bit-by-bit idential with the same hashsum).
My understanding is, that GitHub generates the tarballs for tags on the fly with no guarantees regarding reproducibility.

**Comment by mbiebl** [comment]
And tarballs / assets that have been uploaded as part of a release are actually "stable".

**Comment by evverx** [comment]
Can Debian use tags instead? I thought that systemd kind of forced everyone to be able to deal with signed tags.

(Just to clarify I don't think tarballs should be uploaded and based on a recent PR in the avahi repository (https://github.com/avahi/avahi/pull/850) it seems OpenWrt is fine with tags as well and it was the last place I wasn't sure about in terms of tags).

**Comment by mbiebl** [comment]
The Debian archive is still tarball based.  I can use a tarball that is based on a git(hub) tag, but as said, those tarballs are generated on the fly and afaiu it's not guaranteed that those tarballs are "stable".

**Comment by evverx** [comment]
I turned on the immutable releases feature (https://docs.github.com/en/code-security/concepts/supply-chain-security/immutable-releases) in the avahi repositories and my understanding is that all the assets including the autogenerated tarballs should be immutable. If they weren't immutable that feature would be really weird.

**Comment by mbiebl** [comment]
> I turned on the immutable releases feature (https://docs.github.com/en/code-security/concepts/supply-chain-security/immutable-releases) in the avahi repositories and my understanding is that all the assets including the autogenerated tarballs should be immutable.

Hm, I don't  see that. The website talks about assets attached to a release but I don't think that includes autogenerated tarballs (based on a tag).

**Comment by mbiebl** [comment]
My understanding is, that github will generate tarballs for tags and caches them for a while. But there is no guarantee that those tarballs are reproducible.

**Comment by mbiebl** [comment]
```
Release assets cannot be modified or deleted: All files attached to the release (such as binaries and archives) are protected from modification or deletion.

```
I might be wrong, but an autogenerated tarball from a git tag is not considered an asset.

**Comment by evverx** [comment]
> The website talks about assets attached to a release but I don't think that includes autogenerated tarballs (based on a tag)

I took a look at dfuzzer (where the immutable releases were turned on as well) and the tag itself can be verified:
```sh
$ gh release verify v2.6
Resolved tag v2.6 to sha1:a955a80f7dd20fd7aeffad91da1c495aab5dbbd3
Loaded attestation from GitHub API
✓ Release v2.6 verified!
```
but the autogenerated tarball I downloaded didn't work out
```sh
$ gh release verify-asset v2.6 v2.6.tar.gz 
attestation for v2.6 does not contain subject sha256:c8160841de0fd13619edd0d13a332a0d810f96311067c8d87d0320eb839f0399
```
so it looks like this feature isn't as useful as I thought. It adds immutable tags though but I add rules to protect tags and branches some time ago even without that feature.

**Comment by evverx** [comment]
Actually it is sneakily mentioned in https://docs.github.com/en/code-security/how-tos/secure-your-supply-chain/secure-your-dependencies/verifying-the-integrity-of-a-release
> This command cannot be used to verify the source code zip file or tarball for a release, since these assets are only created when a download is requested.

**Comment by Croydon** [comment]
@mbiebl understanding is correct. GitHub might change how snapshots are generated at any time. They tried to change it once, and it broke an enmours amount of projects relying on the stability of those snapshots https://github.blog/open-source/git/update-on-the-future-stability-of-source-code-archives-and-hashes/

Avahi should generate official release downloads themselves, to have a stable file forever, that can be verified.

Protected tags are repository rules that repository owners can change at any time, without the public knowing. They mostly protect from accidents, and maybe to limit what contributors with less access rights can do.  
On the other hand, when a GitHub release is set as immutable, then GitHub itself (somewhat) guarantees that, so it is more trustworthy. Even when an account got hacked, you can't just swap out the release artifacts at a later date.
